plugin/evm/vm.go

```diff
@@ -496,9 +464,17 @@ func (vm *VM) internalBuildBlock() (chainState.Block, error) {
 	vm.buildBlockLock.Unlock()
 
 	if block == nil {
+		// Signal the mempool that if it was attempting to issue an atomic
+		// transaction into the next block, block building failed and the
+		// transaction should be re-issued unless it was discarded during
+		// atomic tx verification.
+		vm.mempool.CancelCurrentTx()
 		return nil, errCreateBlock
 	}
 
+	// Marks the current tx from the mempool as being successfully issued
+	// into a block.
+	vm.mempool.IssueCurrentTx()
```

Comment:
I need to review more of the async code here to ensure there isn't a race here (where we build the block and then a mempool tx is added).

Reply:
Just pushed a refactor of the bd timer, which should make the logic a bit easier to reason about.

Reply:
Much better! Thanks!